src/OrchardCore/OrchardCore.Queries.Core/QuerySourceHelper.cs

```diff
@@ -0,0 +1,103 @@
+using System.Collections.Generic;
+using System.Text.Json.Nodes;
+using Dapper;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Data;
+using OrchardCore.Environment.Shell.Scope;
+using YesSql;
+using YesSql.Sql;
+
+namespace OrchardCore.Queries.Core;
+
+public static class QuerySourceHelper
+{
+    public static Query CreateQuery(string source, bool canReturnContentItems = true, JsonNode data = null)
+    {
+        var query = new Query()
+        {
+            Source = source,
+            CanReturnContentItems = canReturnContentItems,
+        };
+
+        if (data != null)
+        {
+            query.Name = data[nameof(Query.Name)]?.GetValue<string>();
+            query.Schema = data[nameof(Query.Schema)]?.GetValue<string>();
+
+            // For backward compatibility, we use the key 'ReturnDocuments'.
+            var returnDocuments = data["ReturnDocuments"];
+
+            if (returnDocuments != null)
+            {
+                query.ReturnContentItems = returnDocuments.GetValue<bool>();
+            }
+
+            var returnContentItems = data[nameof(Query.ReturnContentItems)];
+
+            if (returnContentItems != null)
+            {
+                query.ReturnContentItems = returnContentItems.GetValue<bool>();
+            }
+        }
+
+        return query;
+    }
+
+    public static void MigrateQueries(string source)
+    {
+        ShellScope.AddDeferredTask(async scope =>
```

Comment:
Why do this as a deferred task instead of it running directly as part of the migrations it's called from?

Reply:
Because when when these migrations first run, we want to ensure that the new `QueryMigrations` run first to create the index which will lock the table. So defferring this will ensure that this logic happens at the end of the request after the index had been created and the SQL lock have been released which will avoid deadlock.

Reply:
yes I think it does. but it does not prevent the dead lock. I tried it without the deferred-task, but encountered the deadlock. I added documentation.